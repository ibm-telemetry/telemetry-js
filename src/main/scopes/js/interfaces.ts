/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
