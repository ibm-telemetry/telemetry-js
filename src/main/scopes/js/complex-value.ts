/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
