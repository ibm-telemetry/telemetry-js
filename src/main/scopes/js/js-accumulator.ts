/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
