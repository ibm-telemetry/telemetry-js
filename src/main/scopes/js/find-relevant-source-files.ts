/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
