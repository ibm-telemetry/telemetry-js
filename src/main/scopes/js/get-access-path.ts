/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
