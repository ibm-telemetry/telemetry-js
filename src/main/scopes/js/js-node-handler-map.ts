/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import { JsNodeHandlerMap } from '../js/interfaces.js'
import { ImportNodeHandler } from '../js/node-handlers/import-node-handler.js'
import { CallExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/call-expression-node-handler.js'
import { ElementAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/element-access-expression-node-handler.js'
import { IdentifierNodeHandler } from './node-handlers/tokens-and-functions-handlers/identifier-node-handler.js'
import { PropertyAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/property-access-expression-node-handler.js'

/**
 * Maps node kinds to handlers that know how to process them to generate JsFunction and JsToken
 * metrics for the JsScope.
 */
export const jsNodeHandlerMap: JsNodeHandlerMap = {
  [ts.SyntaxKind.ImportDeclaration]: ImportNodeHandler,
  [ts.SyntaxKind.CallExpression]: CallExpressionNodeHandler,
  [ts.SyntaxKind.PropertyAccessExpression]: PropertyAccessExpressionNodeHandler,
  [ts.SyntaxKind.ElementAccessExpression]: ElementAccessExpressionNodeHandler,
  [ts.SyntaxKind.Identifier]: IdentifierNodeHandler
}
