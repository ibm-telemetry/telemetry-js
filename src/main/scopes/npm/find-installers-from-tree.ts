/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { type Logger } from '../../core/log/logger.js'
import { findNestedDeps } from './find-nested-deps.js'
import { getPackageSubTree } from './get-package-sub-tree.js'
import type { DependencyTree } from './interfaces.js'
import { type InstallingPackage } from './interfaces.js'

/*
 * Nested dependency considerations:
 *
 * Given the following hierarchy (assume "b" is an instrumented package):
 *
 * "Root"
 *  └── "a"
 *       └── "b".
 *
 * When "b" gets installed for the first time in the project (as a part of "a"), it will be
 * installed into the same node_modules folder into which "a" was installed (the Root's
 * node_modules folder), even though it is a dependency of "a" and not of the Root project.
 *
 * This nested dependency is not shown in an `npm ls` command run from the Root folder, even
 * though "b" is present in the Root node_modules folder../
 *
 * To solve this, we run `npm ls --all` which will show all installed packages, rather than only
 * those directly depended upon by the Root project.
 */

/**
 * Given a dependency tree, package name, and package version, finds all packages which install the
 * specified package name/version combo. Returns an array of results; or an empty array if no
 * installers were found.
 *
 * @param dependencyTree - The tree to search.
 * @param packageName - The name of the package for which to search.
 * @param filterFn - Function to filter results by.
 * @param logger - A logger instance.
 * @returns An array of results.
 */
export function findInstallersFromTree(
  dependencyTree: DependencyTree,
  packageName: string,
  filterFn: ({ value }: { value: InstallingPackage }) => boolean,
  logger: Logger
) {
  logger.traceEnter('', 'findInstallersFromTree', [dependencyTree, packageName, filterFn])

  let results: InstallingPackage[] = []

  // Matches come back as something like: [..., parentPkgName, dependencies, instrumentedPackage]
  const matches = findNestedDeps(dependencyTree, packageName, filterFn)

  if (matches.length >= 1) {
    // We want to ignore last 2 pieces to get the parent's info, not the child's
    results = matches.map((match) => getPackageSubTree(dependencyTree, match.slice(0, -2)))
  }

  logger.traceExit('', 'findInstallersFromTree', results)
  return results
}
