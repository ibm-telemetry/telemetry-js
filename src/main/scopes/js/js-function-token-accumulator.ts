/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
