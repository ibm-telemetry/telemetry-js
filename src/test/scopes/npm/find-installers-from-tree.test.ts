/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { describe, expect, it } from 'vitest'

import { findInstallersFromTree } from '../../../main/scopes/npm/find-installers-from-tree.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'

const testDependencyTree = await new Fixture('dependency-trees/basic-dependency-tree.json').parse()

describe('findInstallersFromTree', () => {
  const logger = initLogger()

  it('returns empty array if dependency does not exist', () => {
    expect(
      findInstallersFromTree(
        testDependencyTree,
        'not-there',
        ({ value }) => value.version === '1.0.0',
        logger
      )
    ).toStrictEqual([])
  })

  it('returns for a single installer', () => {
    expect(
      findInstallersFromTree(
        testDependencyTree,
        'two',
        ({ value }) => value.version === '1.0.0',
        logger
      )
    ).toMatchSnapshot()
  })

  it('returns for multiple installers', () => {
    expect(
      findInstallersFromTree(
        testDependencyTree,
        'three',
        ({ value }) => value.version === '1.0.0',
        logger
      )
    ).toMatchSnapshot()
  })

  it('only picks up correct version', () => {
    expect(
      findInstallersFromTree(
        testDependencyTree,
        'four',
        ({ value }) => value.version === '1.0.1',
        logger
      )
    ).toMatchSnapshot()
  })

  it('disregards other versions', () => {
    expect(
      findInstallersFromTree(
        testDependencyTree,
        'four',
        ({ value }) => value.version === 'not-there',
        logger
      )
    ).toStrictEqual([])
  })
})
