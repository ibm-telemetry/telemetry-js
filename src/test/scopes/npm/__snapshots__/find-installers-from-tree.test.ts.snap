// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findInstallersFromTree > only picks up correct version 1`] = `
[
  {
    "dependencies": [
      {
        "name": "four",
        "version": "1.0.1",
      },
    ],
    "name": "three",
    "version": "1.0.0",
  },
]
`;

exports[`findInstallersFromTree > returns for a single installer 1`] = `
[
  {
    "dependencies": [
      {
        "name": "two",
        "version": "1.0.0",
      },
      {
        "name": "four",
        "version": "1.0.0",
      },
      {
        "name": "three",
        "version": "1.0.0",
      },
    ],
    "name": "one",
    "version": "1.0.0",
  },
]
`;

exports[`findInstallersFromTree > returns for multiple installers 1`] = `
[
  {
    "dependencies": [
      {
        "name": "two",
        "version": "1.0.0",
      },
      {
        "name": "four",
        "version": "1.0.0",
      },
      {
        "name": "three",
        "version": "1.0.0",
      },
    ],
    "name": "one",
    "version": "1.0.0",
  },
  {
    "dependencies": [
      {
        "name": "three",
        "version": "1.0.0",
      },
    ],
    "name": "two",
    "version": "1.0.0",
  },
]
`;
